File: rust/crates/parity/sdk/parity_sdk.py
#!/usr/bin/env python3
"""Official-SDK parity check (Python `azure-cosmos`).

This drives a *running* Rust emulator with the real Azure Cosmos DB Python SDK,
proving that an official client can speak to the port unchanged. It is an
**opt-in** layer: it requires network access to install the SDK and is NOT run by
`cargo test`.

Usage:
    # 1. Start the emulator (foreground), noting the key it prints:
    cargo run -p cosmos-cli -- start --key <master-key>
    # 2. In another shell:
    pip install azure-cosmos
    python3 crates/parity/sdk/parity_sdk.py \
        --endpoint http://localhost:8081 --key <master-key>

The default endpoint/key match the emulator's well-known development defaults.
"""
from __future__ import annotations

import argparse
import sys
import uuid

DEFAULT_KEY = (
    "C2y6yDjf5/R+ob0N8A7Cgv30VRDJIWEHLM+4QDU5DE2nQ9nDuVTqobD4b8mGGyPMbIZ"
    "nqyMsEcaGQy67XIw/Jw=="
)


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--endpoint", default="http://localhost:8081")
    parser.add_argument("--key", default=DEFAULT_KEY)
    args = parser.parse_args()

    try:
        from azure.cosmos import CosmosClient, PartitionKey
    except ImportError:
        print("azure-cosmos is not installed. Run: pip install azure-cosmos", file=sys.stderr)
        return 2

    client = CosmosClient(args.endpoint, credential=args.key)

    db_id = f"db-{uuid.uuid4().hex}"
    coll_id = f"coll-{uuid.uuid4().hex}"

    db = client.create_database(db_id)
    print(f"[ok] created database {db_id}")

    container = db.create_container(id=coll_id, partition_key=PartitionKey(path="/partitionKey"))
    print(f"[ok] created container {coll_id}")

    doc_id = uuid.uuid4().hex
    created = container.create_item({"id": doc_id, "partitionKey": "tenant-1", "value": "created"})
    assert created["value"] == "created"
    print(f"[ok] created document {doc_id}")

    read = container.read_item(item=doc_id, partition_key="tenant-1")
    assert read["value"] == "created"
    print("[ok] read document")

    read["value"] = "updated"
    replaced = container.replace_item(item=doc_id, body=read)
    assert replaced["value"] == "updated"
    print("[ok] replaced document")

    container.delete_item(item=doc_id, partition_key="tenant-1")
    print("[ok] deleted document")

    client.delete_database(db_id)
    print("[ok] deleted database")

    print("PARITY_SDK_PYTHON_OK")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: rust/crates/parity/sdk/parity_sdk.js
#!/usr/bin/env node
/*
 * Official-SDK parity check (Node `@azure/cosmos`).
 *
 * Drives a *running* Rust emulator with the real Azure Cosmos DB JavaScript SDK.
 * Opt-in layer: requires network access to `npm install @azure/cosmos` and is
 * NOT run by `cargo test`.
 *
 * Usage:
 *   cargo run -p cosmos-cli -- start --key <master-key>
 *   npm install @azure/cosmos
 *   node crates/parity/sdk/parity_sdk.js --endpoint http://localhost:8081 --key <master-key>
 */
"use strict";

const DEFAULT_KEY =
  "C2y6yDjf5/R+ob0N8A7Cgv30VRDJIWEHLM+4QDU5DE2nQ9nDuVTqobD4b8mGGyPMbIZnqyMsEcaGQy67XIw/Jw==";

function arg(name, fallback) {
  const i = process.argv.indexOf(name);
  return i >= 0 && i + 1 < process.argv.length ? process.argv[i + 1] : fallback;
}

async function main() {
  let CosmosClient;
  try {
    ({ CosmosClient } = require("@azure/cosmos"));
  } catch (e) {
    console.error("@azure/cosmos is not installed. Run: npm install @azure/cosmos");
    process.exit(2);
  }

  const endpoint = arg("--endpoint", "http://localhost:8081");
  const key = arg("--key", DEFAULT_KEY);
  const client = new CosmosClient({ endpoint, key });

  const dbId = `db-${Date.now().toString(16)}`;
  const collId = `coll-${Date.now().toString(16)}`;

  const { database } = await client.databases.create({ id: dbId });
  console.log(`[ok] created database ${dbId}`);

  const { container } = await database.containers.create({
    id: collId,
    partitionKey: { paths: ["/partitionKey"] },
  });
  console.log(`[ok] created container ${collId}`);

  const docId = Math.random().toString(16).slice(2);
  const { resource: created } = await container.items.create({
    id: docId,
    partitionKey: "tenant-1",
    value: "created",
  });
  if (created.value !== "created") throw new Error("create mismatch");
  console.log(`[ok] created document ${docId}`);

  const { resource: read } = await container.item(docId, "tenant-1").read();
  if (read.value !== "created") throw new Error("read mismatch");
  console.log("[ok] read document");

  read.value = "updated";
  const { resource: replaced } = await container.item(docId, "tenant-1").replace(read);
  if (replaced.value !== "updated") throw new Error("replace mismatch");
  console.log("[ok] replaced document");

  await container.item(docId, "tenant-1").delete();
  console.log("[ok] deleted document");

  await database.delete();
  console.log("[ok] deleted database");

  console.log("PARITY_SDK_NODE_OK");
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
